blog/_posts/2022-01-01-apis-under-control.md

```diff
@@ -0,0 +1,110 @@
+---
+slug: "apis-under-control"
+date: 2022-01-01
+author: siegs
+layout: blog
+title: "APIs Under Control"
+subtitle: "Why the Kubernetes APIs are so powerful"
+description: "We explore what makes the Kubernetes approach to APIs so useful, and how we applied them to Tilt."
+image: "/assets/images/apis-under-control/flying-jeep.jpg"
+tags:
+  - kubernetes
+  - api
+  - api-server
+---
+An underrated aspect of Kubernetes is how open and extensible the system is. Every facet of a Kubernetes cluster can be browsed, inspected and modified via a standard [REST API][K8SREST]. [Custom resources][CustomResources] can be defined in the API to hold data specific to each production system. [Operators][] can be added to react to changes to the system and perform new functionality that goes beyond what the Kubernetes developers originally intended.
+
+At its core, Kubernetes is more than an API or even a container orchestration platform, it's a pattern for building distributed systems, modeled after a [feedback control system][control-system] from systems engineering. In essence:
+
+1. Declare a desired _setpoint_ for the system, using a [uniform interface][REST] (resources in the Kubernetes API).
+2. Operate a _[control loop][control-loop]_ to monitor all inputs to the system and converge the system state towards the setpoint (the Kubernetes [control plane][control-plane]).
+3. Allow any interested consumer to monitor inputs and make changes to the system and its setpoint ([watches][] and [operators][Operators]).
+
+In this way, Kubernetes does for computing workloads what cruise control (speed as a setpoint) or autopilot (destination as a setpoint) do for your automobile, except in an open and extensible fashion. (This reminds me of a [vulnerability in Fiat/Chrysler/Jeep automobiles][jeep] years ago that allowed remote attackers complete control of the vehicle. [Secure your clusters][secure-clusters], everyone!)
```

Comment:
This is an important point, let's spend a bit more time on this?
Why is this so awesome? Some ideas:
- Setup & management of a k8s cluster can be completely automated and reproduced, making scaling up fast and convenient
- self-healing / keeping cluster states consistent is easy to implement / manage